### ISSUE TITLE
readNonBlocking() and readLatest() don't always call postRead()

### ISSUE THREAD

**Issue by killenb** [ChimeraTK/DeviceAccess#116]
Child of ChimeraTK/ApplicationCore#123

readNonBlocking() and readLatest() only call postRead if there was new data. This is wrong. preRead() and postRead() must always be called in pairs, for instance if the user buffer needs to be restored.
In addidion, this will prevent proper exception propagation with the new exception handling scheme.

Solution in order not to perform unnecessary calculations:
change postRead() and doPostRead() to postRead(bool hasNewData) and doPostRead(bool hasNewData), and let the implementations decide what to do in case of no data.

* [x] Introduce the API change to postRead() and doPostRead() in all places
* [x] In read() and TransferFuture::wait() and transferGroup::read(), call postRead(true) after a successful read transfer, and postRead(false) if there was an exception
* [x] In readLatest() and readNonBlocking() call it with the return value of the corresponding doReadXxx if there was no exception, and false if there was an exception
* [x] Change postWrite to receive a data lost flag (see comment below)

* Do not change any implementations yet.
* [x] All tests should still pass. If not, note in this ticket which tests are failing.
* [x] If there are failing test check if they belong to the follow-up ticket #122. Note in that ticket which corresponding tests are failing.
* [x] Fix all remaining tests (or the implementations such that the tests pass, depending on which of the two does not correspond to the new spec).
* [x] Merge into the master

**Comment by killenb** [comment]
Fixed ApplicationCore ExceptionHandling Spec 11.23

**Comment by killenb** [comment]
The "hasNewData" flag is the return value of "readTransferNonBlocking" or "readTransferLatest". For readTransfer use true.

For symmetry reasons the return value of write shall be added as postWrite(bool dataLost).

**Comment by killenb** [comment]
Do before of after #119. It introduces an interface change at the same place.

**Comment by ckampm** [comment]
> The "hasNewData" flag is the return value of "readTransferNonBlocking" or "readTransferLatest". For readTransfer use true.

This statement is contradicting with the second task in the opening comment. After `readTransfer`, the `hasSeenException` flag has to be tested in order to check if the transfer has been successful and provided new data.

**Comment by ckampm** [comment]
,
